src/core/aio.c

```diff
@@ -505,34 +508,57 @@ nni_aio_list_active(nni_aio *aio)
 static void
 nni_aio_expire_add(nni_aio *aio)
 {
-	nni_list *list = &aio->a_expire_q->eq_list;
-	nni_aio * prev;
+	nni_aio_expire_q * eq = aio->a_expire_q;
+	if (eq->eq_len > eq->eq_cap) {
+		nni_aio ** new_list =
+		    nni_zalloc(eq->eq_cap * 2 * sizeof(nni_aio *));
+		for (uint32_t i = 0; i < eq->eq_len; i++) {
+			new_list[i] = eq->eq_list[i];
+		}
+		nni_free(eq->eq_list, eq->eq_cap * sizeof(nni_aio *));
+		eq->eq_list = new_list;
+		eq->eq_cap *= 2;
+	}
+	eq->eq_list[eq->eq_len++] = aio;
+	if (eq->eq_len == 1) {
+		nni_cv_wake(&eq->eq_cv);
+	}
+}
 
-	// This is a reverse walk of the list.  We're more likely to find
-	// a match at the end of the list.
-	for (prev = nni_list_last(list); prev != NULL;
-	     prev = nni_list_prev(list, prev)) {
-		if (aio->a_expire >= prev->a_expire) {
-			nni_list_insert_after(list, aio, prev);
+static void
+nni_aio_expire_rm(nni_aio *aio)
+{
+	uint32_t          aio_idx = 0;
+	nni_aio_expire_q *eq      = aio->a_expire_q;
+	for (aio_idx = 0; aio_idx < eq->eq_len; aio_idx++) {
+		if (aio == eq->eq_list[aio_idx]) {
+			eq->eq_list[aio_idx] = eq->eq_list[eq->eq_len-1];
+			eq->eq_len --;
 			break;
 		}
 	}
-	if (prev == NULL) {
-		// This has the shortest time, so insert at the start.
-		nni_list_prepend(list, aio);
-		// And, as we are the latest, kick the thing.
-		nni_cv_wake(&aio->a_expire_q->eq_cv);
+
+	if (eq->eq_len < eq->eq_cap / 3 && eq->eq_cap > NNI_EXPIRE_Q_SIZE) {
```

Comment:
Please consider changing / 3 to / 4.  Division by 3 is very expensive.  Division by 4 (or any other power of 2) is super cheap.  This might make it slower to reclaim memory, but really, that's probably ok.  If we bursted up to this value, we might burst again, and using a quarter as the low threshold to free should further reduce pressure on the allocator for bursty workloads.